### PR TITLE
Azure monitor query

### PR DESCRIPTION
**Pull request by rakshith91**

### CODE REVIEW

File: sdk/monitor/azure-monitor-query/azure/monitor/query/aio/_log_query_client_async.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
+#
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+from typing import Any, Union, Sequence, Dict
+from azure.core.exceptions import HttpResponseError
+from azure.core.credentials_async import AsyncTokenCredential
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/aio/_log_query_client_async.py:10]:
```suggestion
try:
    from azure.core.credentials_async import AsyncTokenCredential
except ImportError:
    pass
```

File: sdk/monitor/azure-monitor-query/azure/monitor/query/aio/_metrics_query_client_async.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+from typing import TYPE_CHECKING, Any
+
+from .._generated.aio._monitor_query_client import (
+    MonitorQueryClient,
+)
+from .._models import MetricsResult, MetricDefinition
+from .._helpers import get_authentication_policy
+
+if TYPE_CHECKING:
+    from azure.core.credentials import TokenCredential
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/aio/_metrics_query_client_async.py:19]:
```suggestion


<!--StartFragment-->try:    from azure.core.credentials_async import AsyncTokenCredentialexcept ImportError:    pass<!--EndFragment-->


```

File: sdk/monitor/azure-monitor-query/azure/monitor/query/aio/_log_query_client_async.py (hunk 1 above)
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/aio/_log_query_client_async.py:10]:
```suggestion
from typing import Any, Union, Sequence, Dict, TYPE_CHECKING
from azure.core.exceptions import HttpResponseError
if TYPE_CHECKING:
    try:
        from azure.core.credentials_async import AsyncTokenCredential
    except ImportError:
        pass
```

File: sdk/monitor/azure-monitor-query/README.md [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+from azure.identity import ClientSecretCredential
+
+
+credential  = ClientSecretCredential(
+        client_id = os.environ['AZURE_CLIENT_ID'],
+        client_secret = os.environ['AZURE_CLIENT_SECRET'],
+        tenant_id = os.environ['AZURE_TENANT_ID']
+    )
+
+client = LogsClient(credential)
```
Comment (srnagar) [inline sdk/monitor/azure-monitor-query/README.md:37]: This should be `LogsQueryClient`.

File: sdk/monitor/azure-monitor-query/README.md [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+from azure.identity import ClientSecretCredential
+
+
+credential  = ClientSecretCredential(
+        client_id = os.environ['AZURE_CLIENT_ID'],
+        client_secret = os.environ['AZURE_CLIENT_SECRET'],
+        tenant_id = os.environ['AZURE_TENANT_ID']
+    )
+
+client = MetricsClient(credential)
```
Comment (srnagar) [inline sdk/monitor/azure-monitor-query/README.md:51]: This should be `MetricsQueryClient`.

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+from ._generated.models import BatchRequest
+from ._helpers import get_authentication_policy, process_error
+from ._models import LogsQueryResults, LogsQueryRequest, LogsQueryBody, LogsBatchResults
+
+if TYPE_CHECKING:
+    from azure.core.credentials import TokenCredential
+
+
+class LogsClient(object):
```
Comment (srnagar) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py:21]: Change the name to `LogsQueryClient`

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    """LogsClient
+
+    :param credential: The credential to authenticate the client
+    :type credential: ~azure.core.credentials.TokenCredential
+    :keyword endpoint: The endpoint to connect to. Defaults to 'https://api.loganalytics.io/v1'.
```
Comment (srnagar) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py:26]:
```suggestion
    :keyword endpoint: The endpoint to connect to. Defaults to 'https://api.loganalytics.io/'.
```

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+        Executes an Analytics query for data.
+
+        :param workspace_id: ID of the workspace. This is Workspace ID from the Properties blade in the
+         Azure portal.
+        :type workspace_id: str
+        :param query: The Analytics query. Learn more about the `Analytics query syntax
+         <https://azure.microsoft.com/documentation/articles/app-insights-analytics-reference/>`_.
+        :type query: str
+        :keyword str timespan: Optional. The timespan over which to query data. This is an ISO8601 time
```
Comment (srnagar) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py:54]: Make timespan a positional param.

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                **kwargs
+            ))
+        except HttpResponseError as e:
+            process_error(e)
+
+    def batch_query(self, queries, **kwargs):
+        # type: (Union[Sequence[Dict], Sequence[LogsQueryRequest]], Any) -> LogsBatchResults
+        """Execute an Analytics query.
+
+        Executes an Analytics query for data.
```
Comment (srnagar) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py:112]: Documentation should be updated for `batch_query` API.

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_metrics_query_client.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+from ._models import MetricsResult, MetricDefinition, MetricNamespace
+from ._helpers import get_metrics_authentication_policy
+
+if TYPE_CHECKING:
+    from azure.core.credentials import TokenCredential
+    from azure.core.paging import ItemPaged
+    from ._models import MetricNamespace
+
+
+class MetricsClient(object):
```
Comment (srnagar) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_metrics_query_client.py:25]: Change name to `MetricsQueryClient`

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_metrics_query_client.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+    :param credential: The credential to authenticate the client
+    :type credential: ~azure.core.credentials.TokenCredential
+    """
+
+    def __init__(self, credential, **kwargs):
+        # type: (TokenCredential, Any) -> None
+        self._client = MonitorQueryClient(
+            credential=credential,
+            base_url='https://management.azure.com',
```
Comment (srnagar) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_metrics_query_client.py:36]: Make endpoint configurable

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        :paramtype qualified_names: list[str]
+        :keyword workspace_ids: A list of workspace IDs that are included in the query.
+        :paramtype workspace_ids: list[str]
+        :keyword azure_resource_ids: A list of Azure resource IDs that are included in the query.
+        :paramtype azure_resource_ids: list[str]
+        :return: QueryResults, or the result of cls(response)
+        :rtype: ~azure.monitor.query.LogsQueryResults
+        :raises: ~azure.core.exceptions.HttpResponseError
+        """
+        timespan = kwargs.pop("timespan", None)
```
Comment (srnagar) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py:75]: Is it easy for the user to construct the ISO 8601 formatted time intervals? Is it better to provide a simpler way to specify the start and end (datetime/duration)?

```
Start and end, such as "2007-03-01T13:00:00Z/2008-05-11T15:30:00Z"
Start and duration, such as "2007-03-01T13:00:00Z/P1Y2M10DT2H30M"
Duration and end, such as "P1Y2M10DT2H30M/2008-05-11T15:30:00Z"
Duration only, such as "P1Y2M10DT2H30M", with additional context information
```

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_models.py [hunk 12]
```diff
@@ -0,0 +1,598 @@
...
+            cost=generated.cost,
+            timespan=generated.timespan,
+            interval=generated.interval,
+            namespace=generated.namespace,
+            resourceregion=generated.resourceregion,
+            metrics=[Metric._from_generated(m) for m in generated.value] # pylint: disable=protected-access
+        )
+
+class LogsQueryRequest(InternalLogQueryRequest):
+    """An single request in a batch.
```
Comment (srnagar) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_models.py:150]:
```suggestion
    """A single request in a batch.
```

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py (hunk 11 above)
Comment (rakshith91) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py:75]: removed timespan kwarg to replace it with start_time, end_time and duration - we'll magically construct the timespan for the users now

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+                **kwargs
+            ))
+        except HttpResponseError as e:
+            process_error(e)
+
+    def batch_query(self, queries, **kwargs):
+        # type: (Union[Sequence[Dict], Sequence[LogsQueryRequest]], Any) -> LogsBatchResults
+        """Execute an Analytics query.
+
+        Executes an Analytics query for data.
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py:119]:
```suggestion
        """Execute a list of analytics queries. Each request can be either a LogQueryRequest
        object or an equivalent serialized model.

        The response is returned in the same order as that of the requests sent.
```

**Review by srnagar — APPROVED** [review]

File: sdk/monitor/azure-monitor-query/README.md [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+]
+response = client.batch_query(requests)
+
+for response in response.responses:
+    body = response.body
+    if not body.tables:
+        print("Something is wrong")
+    else:
+        for table in body.tables:
+            df = pd.DataFrame(table.rows, columns=[col.name for col in table.columns])
```
[the remark this answers is not in the document]
Comment (rakshith91) [reply sdk/monitor/azure-monitor-query/README.md:185]: yes - the sample has both with and without pandas as an example

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+        :param query: The Analytics query. Learn more about the `Analytics query syntax
+         <https://azure.microsoft.com/documentation/articles/app-insights-analytics-reference/>`_.
+        :type query: str
+        :keyword datetime start_time: The start time from which to query the data. This should be accompanied
+         with either end_time or duration.
+        :keyword datetime end_time: The end time till which to query the data. This should be accompanied
+         with either start_time or duration.
+        :keyword str duration: The duration for which to query the data. This can also be accompanied
+         with either start_time or end_time. If start_time or end_time is not provided, the current time is
+         taken as the end time. This should be provided in a ISO8601 string format like 'PT1H', 'P1Y2M10DT2H30M'.
```
[the remark this answers is not in the document]
Comment (rakshith91) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py:60]: sounds good - for beta 1, i'll make duration positional parameter with optional start and end time

File: sdk/monitor/azure-monitor-query/tests/test_metrics_client.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+def _credential():
+    credential  = ClientSecretCredential(
+        client_id = os.environ['AZURE_CLIENT_ID'],
+        client_secret = os.environ['AZURE_CLIENT_SECRET'],
+        tenant_id = os.environ['AZURE_TENANT_ID']
+    )
+    return credential
+
+@pytest.mark.liveTest
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/tests/test_metrics_client.py:14]:
```suggestion
@pytest.mark.live_test_only
```

File: sdk/monitor/azure-monitor-query/tests/test_metrics_client.py [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+def test_metrics_auth():
+    credential = _credential()
+    client = MetricsQueryClient(credential)
+    # returns LogsQueryResults 
+    response = client.query(os.environ['METRICS_RESOURCE_URI'], metric_names=["PublishSuccessCount"], timespan='P2D')
+
+    assert response is not None
+    assert response.metrics is not None
+
+@pytest.mark.liveTest
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/tests/test_metrics_client.py:24]:
```suggestion
@pytest.mark.live_test_only
```

File: sdk/monitor/azure-monitor-query/tests/test_metrics_client.py [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+def test_metrics_namespaces():
+    client = MetricsQueryClient(_credential())
+
+    response = client.list_metric_namespaces(os.environ['METRICS_RESOURCE_URI'])
+
+    assert response is not None
+
+@pytest.mark.liveTest
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/tests/test_metrics_client.py:32]:
```suggestion
@pytest.mark.live_test_only
```

File: sdk/monitor/azure-monitor-query/tests/conftest.py [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+#
+# --------------------------------------------------------------------------
+import platform
+import sys
+
+
+# Ignore async tests for Python < 3.5
+collect_ignore_glob = []
+if sys.version_info < (3, 5):
+    collect_ignore_glob.append("*_async.py")
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/tests/conftest.py:33]:
```suggestion
    collect_ignore_glob.append("*.py")
```

File: sdk/monitor/azure-monitor-query/tests/conftest.py [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+#
+# --------------------------------------------------------------------------
+import platform
+import sys
+
+
+# Ignore async tests for Python < 3.5
+collect_ignore_glob = []
+if sys.version_info < (3, 5):
+    collect_ignore_glob.append("*.py")
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/tests/conftest.py:33]:
```suggestion
    collect_ignore_glob.append("*_async.py")
```

File: sdk/monitor/azure-monitor-query/setup.py [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+        # Exclude packages that will be covered by PEP420 or nspkg
+        'azure',
+        'azure.monitor',
+    ]),
+    install_requires=[
+        'msrest>=0.6.19',
+        'azure-core<2.0.0,>=1.12.0',
+    ],
+    extras_require={
+        ":python_version<'3.0'": ['azure-nspkg'],
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/setup.py:88]:
```suggestion
        ":python_version<'3.0'": ['azure-monitor-nspkg'],
```

File: sdk/monitor/azure-monitor-query/samples/async_samples/sample_metric_definitions_async.py [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+async def list_namespaces():
+    credential  = ClientSecretCredential(
+            client_id = os.environ['AZURE_CLIENT_ID'],
+            client_secret = os.environ['AZURE_CLIENT_SECRET'],
+            tenant_id = os.environ['AZURE_TENANT_ID']
+        )
+
+    client = MetricsQueryClient(credential)
+
+    metrics_uri = "/subscriptions/faa080af-c1d8-40ad-9cce-e1a450ca5b57/resourceGroups/sabhyrav-resourcegroup/providers/Microsoft.EventGrid/topics/rakshith-cloud"
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/samples/async_samples/sample_metric_definitions_async.py:18]:
```suggestion
    metrics_uri = os.environ['METRICS_RESOURCE_URI']
```

**Pull request merged**